snap/info.go

```diff
@@ -950,6 +951,109 @@ func (s *Info) DesktopPrefix() string {
 	return fmt.Sprintf("%s+%s", s.SnapName(), s.InstanceKey)
 }
 
+// DesktopFileIDs returns desktop-file-ids desktop plug attribute entries.
+// The desktop-file-ids attribute is optional so an empty list is returned if
+// the it is not found.
+//
+// Note: DesktopFileIDs doesn't check if the desktop plug is connected because
+// the desktop-file-ids attribute is controlled by an allow-installation rule.
+func (s *Info) DesktopFileIDs() ([]string, error) {
+	var desktopPlug *PlugInfo
+	for _, plug := range s.Plugs {
+		if plug.Interface == "desktop" {
+			desktopPlug = plug
+			break
+		}
+	}
+	if desktopPlug == nil {
+		return nil, nil
+	}
+
+	attrVal, exists := desktopPlug.Lookup("desktop-file-ids")
+	if !exists {
+		// desktop-file-ids attribute is optional
+		return nil, nil
+	}
+
+	// TODO: The internal errors below should never happen due to validation
+	// in the desktop interface. It would be a good candidate for telemetry
+	// error reporting.
+
+	// desktop-file-ids must be a list of strings
+	attrList, ok := attrVal.([]interface{})
+	if !ok {
+		return nil, errors.New(`internal error: "desktop-file-ids" must be a list of strings`)
+	}
+
+	desktopFileIDs := make([]string, 0, len(attrList))
+	for _, val := range attrList {
+		desktopFileID, ok := val.(string)
+		if !ok {
+			return nil, errors.New(`internal error: "desktop-file-ids" must be a list of strings`)
+		}
+		desktopFileIDs = append(desktopFileIDs, desktopFileID)
+	}
+	return desktopFileIDs, nil
+}
+
+// MangleDesktopFileName returns the file name prefixed with Info.DesktopPrefix()
+// unless its name (without the .desktop extension) is listed under the
+// desktop-file-ids desktop interface attribute.
+func (s *Info) MangleDesktopFileName(desktopFile string) (string, error) {
```

Comment:
does this need to be exported?

Reply:
Yes, I was trying to have the mangling consistent between the desktop wrappers and apparmor generation to avoid any drifts in implementation between the two.

Desktop wrappers, need to open the snap's desktop files under `<snap-mount>/meta/gui`, do some processing and write them (with the mangled names) under `SnapDesktopFilesDir`. [deriveDesktopFilesContent](https://github.com/canonical/snapd/pull/14456/files#diff-2ad17151ef3d48258927d3cf77db183624b95593d8c195db498a11d5d56df8acR250)

Apparmor rules generation (https://github.com/canonical/snapd/pull/14444 which builds on top of this PR) only needs the installed mangled names but since generation profiles happens after mounting but before installing the desktop files, we need to infer them using `DesktopFilesFromMount` with `MangleFileNames=true`. [getDesktopFileRules](https://github.com/canonical/snapd/pull/14444/files#diff-d004a399bf1eea823f934a75d258a3388a825af02f2688a34997a81e53ce14beR110)

I would appreciate any suggestions to make it simpler.